flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java

```diff
@@ -0,0 +1,209 @@
+	+ "hohen Werke Sind herrlich wie am ersten Tag.\n"
+	+ "GABRIEL: Und schnell und unbegreiflich schnelle Dreht sich umher der Erde\n"
+	+ "Pracht; Es wechselt Paradieseshelle Mit tiefer, schauervoller Nacht. Es\n"
+	+ "schaeumt das Meer in breiten Fluessen Am tiefen Grund der Felsen auf, Und\n"
+	+ "Fels und Meer wird fortgerissen Im ewig schnellem Sphaerenlauf.\n"
+	+ "MICHAEL: Und Stuerme brausen um die Wette Vom Meer aufs Land, vom Land\n"
+	+ "aufs Meer, und bilden wuetend eine Kette Der tiefsten Wirkung rings umher.\n"
+	+ "Da flammt ein blitzendes Verheeren Dem Pfade vor des Donnerschlags. Doch\n"
+	+ "deine Boten, Herr, verehren Das sanfte Wandeln deines Tags.";
+
+	@Rule
+	public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+	private FileCache fileCache;
+
+	private final PermanentBlobKey permanentBlobKey = new PermanentBlobKey();
+
+	private final PermanentBlobService blobService = new PermanentBlobService() {
+		@Override
+		public File getFile(JobID jobId, PermanentBlobKey key) throws IOException {
+			if (key.equals(permanentBlobKey)) {
+				final File zipArchive = temporaryFolder.newFile("zipArchive");
+				try (ZipOutputStream zis = new ZipOutputStream(new FileOutputStream(zipArchive))) {
+
+					final ZipEntry zipEntry = new ZipEntry("cacheFile");
+					zis.putNextEntry(zipEntry);
+
+					IOUtils.copyBytes(new ByteArrayInputStream(testFileContent.getBytes(StandardCharsets.UTF_8)), zis, false);
+				}
+				return zipArchive;
+			} else {
+				throw new IllegalArgumentException("This service contains only entry for " + permanentBlobKey);
+			}
+		}
+
+		@Override
+		public void close() throws IOException {
+
+		}
+	};
+
+	@Before
+	public void setup() throws Exception {
+		fileCache = new FileCache(new String[]{temporaryFolder.newFolder().getAbsolutePath()}, blobService);
+	}
+
+	@After
+	public void shutdown() {
+		fileCache.shutdown();
+	}
+
+	@Test
+	public void testDirectoryDownloadedFromBlob() throws Exception {
+		JobID jobID = new JobID();
+		ExecutionAttemptID attemptID = new ExecutionAttemptID();
+
+		final String fileName = "test_file";
+		// copy / create the file
+		final DistributedCache.DistributedCacheEntry entry = new DistributedCache.DistributedCacheEntry(
+			fileName,
+			false,
+			InstantiationUtil.serializeObject(permanentBlobKey),
+			true);
+		Future<Path> copyResult = fileCache.createTmpFile(fileName, entry, jobID, attemptID);
+
+		final Path dstPath = copyResult.get();
+		final FileSystem fs = dstPath.getFileSystem();
+		final FileStatus fileStatus = fs.getFileStatus(dstPath);
+		assertTrue(fileStatus.isDir());
+
+		final Path cacheFile = new Path(dstPath, "cacheFile");
+		assertTrue(fs.exists(cacheFile));
+		final String actualContent = FileUtils.readFileUtf8(new File(cacheFile.getPath()));
+		assertEquals(testFileContent, actualContent);
+	}
+
+	@Test
+	public void testDirectoryCleanUp() throws Exception {
+		fileCache.shutdown();
```

Comment:
Could we not always pass a `DeleteCapturingDirectScheduledExecutorService` into the FileCache? We can explicitly call run in `@After`; that would allow us to share the filecache between both tests again.